### ISSUE TITLE
tf.estimator.train_and_evaluate auc value is different from tf.estimator.evaluate with same model

### ISSUE THREAD

**Issue by geyuxiao** [tensorflow/tensorflow#22857]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: no
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:no 
- **TensorFlow installed from (source or binary)**:
- **TensorFlow version (use command below)**: 'v1.11.0-0-gc19e29306c', '1.11.0'
- **Python version**:2.7
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

### Describe the problem

I use tf.Estimator for training my models.
Here is part of my source code:

### Source code/log
**tf.estimator.train_and_evaluate**

```
 for n in range(num_epochs):
    train_spec = tf.estimator.TrainSpec(input_fn=lambda: input_fn(...))
    eval_spec = tf.estimator.EvalSpec(input_fn =lambda: input_fn(...))
    tf.estimator.train_and_evaluate(model, train_spec, eval_spec)


INFO:tensorflow:Saving dict for global step 18621: auc = 0.6128701, global_step = 18621, loss = 0.17829555
INFO:tensorflow:Saving dict for global step 37242: auc = 0.6212857, global_step = 37242, loss = 0.17826068
INFO:tensorflow:Saving dict for global step 55863: auc = 0.6261746, global_step = 55863, loss = 0.17617485
INFO:tensorflow:Saving dict for global step 74484: auc = 0.630533, global_step = 74484, loss = 0.17646796
INFO:tensorflow:Saving dict for global step 93105: auc = 0.63444453, global_step = 93105, loss = 0.17552255
INFO:tensorflow:Saving dict for global step 111726: auc = 0.63371396, global_step = 111726, loss = 0.17790207
INFO:tensorflow:Saving dict for global step 130347: auc = 0.62660295, global_step = 130347, loss = 0.17602158
INFO:tensorflow:Saving dict for global step 148968: auc = 0.6290031, global_step = 148968, loss = 0.17708226
INFO:tensorflow:Saving dict for global step 167589: auc = 0.6234657, global_step = 167589, loss = 0.17724389
INFO:tensorflow:Saving dict for global step 186210: auc = 0.62800914, global_step = 186210, loss = 0.17702612
```

### Source code/log

**tf.estimator.evaluate**

```
ckpts = [18621, 37242, 55863, 74484, 93105, 111726, 130347, 148968, 167589, 186210]
        for epoch, ckpt in enumerate(ckpts):
            model.evaluate(
                input_fn=lambda: input_fn(...),
                checkpoint_path=FLAGS.model_dir + "/model.ckpt-" + str(ckpt))

INFO:tensorflow:Saving dict for global step 18621: auc = 0.6088399, global_step = 18621, loss = 0.1782806
INFO:tensorflow:Saving dict for global step 37242: auc = 0.61857945, global_step = 37242, loss = 0.17766058
INFO:tensorflow:Saving dict for global step 55863: auc = 0.6241194, global_step = 55863, loss = 0.17692827
INFO:tensorflow:Saving dict for global step 74484: auc = 0.627106, global_step = 74484, loss = 0.17647153
INFO:tensorflow:Saving dict for global step 93105: auc = 0.62898576, global_step = 93105, loss = 0.17620046
INFO:tensorflow:Saving dict for global step 111726: auc = 0.62996054, global_step = 111726, loss = 0.17710653
INFO:tensorflow:Saving dict for global step 130347: auc = 0.63045925, global_step = 130347, loss = 0.17653853
INFO:tensorflow:Saving dict for global step 148968: auc = 0.6306327, global_step = 148968, loss = 0.17709586
INFO:tensorflow:Saving dict for global step 167589: auc = 0.6306394, global_step = 167589, loss = 0.1774284
INFO:tensorflow:Saving dict for global step 186210: auc = 0.6304847, global_step = 186210, loss = 0.17686236
```

my question is why same model but different auc value. To validate which is right. I also try auc api by sklearn.The result is that auc value same with tf.estimator.evaluate 

### Source code/log
```
 ckpts = [18621, 37242, 55863, 74484, 93105, 111726, 130347, 148968, 167589, 186210]
        for epoch, ckpt in enumerate(ckpts):
            for data_from, data_mode in zip([FLAGS.valid_data], ['test']):
                preds = model.predict(
                    input_fn=lambda: input_fn.(...),
                    checkpoint_path=FLAGS.model_dir + "/model.ckpt-" + str(ckpt))
                label_list = []
                pred_list = []
                for i, prob in enumerate(preds):
                    pred_list.append(prob['prob'][0])
                    label_list.append(prob['true_label'])
                auc_local = sklearn.metrics.roc_auc_score(label_list, pred_list)
                print("%s sklearn.auc is :", (epoch, data_mode, auc_local))


%s auc is : (0, 'test', 0.608910481886984)
%s auc is : (1, 'test', 0.6186364169158124)
%s auc is : (2, 'test', 0.6242379534016801)
%s auc is : (3, 'test', 0.6272265511554851)
%s auc is : (4, 'test', 0.6291364474305191)
%s auc is : (5, 'test', 0.6300167648655547)
%s auc is : (6, 'test', 0.6305495577793957)
%s auc is : (7, 'test', 0.6307263729168348)
%s auc is : (8, 'test', 0.6307389062228584)
%s auc is : (9, 'test', 0.6306225969185617)
```

**Comment by geyuxiao** [comment]
@wt-huang .I suspect this is a bug? Or is there a problem with my use? Could you give me some advice~~thanks a lot .

**Comment by wt-huang** [comment]
@geyuxiao `tf.estimator.train_and_evaluate` does evaluation during training and is essentially alternating between `tf.estimator.train` and `tf.estimator.evalute`. However the ckpts in `tf.estimator.train`_and `tf.estimator.evalute` may have a slight delay in reporting. 

In this case, use tf.estimator.evalute and the corresponding auc values if you only care about final evaluation results.

**Comment by geyuxiao** [comment]
Thank you very much for your reply. You mentioned that there will be a slight delay in reporting. What is the reason for this delay? Or how to avoid this delay. I hope to get a further answer. Because for me, train_and_evaluate is more efficient, and then evaluate after the training. Look forward to your reply.

**Comment by tracihuang9** [comment]
@geyuxiao The delay between tf.estimator.train_and tf.estimator.evalute depends on your model and datasets, among other factors. 
The delay is usually very small so the difference in loss should be minimal and is unlikely to have statistical and practical significance.

**Comment by wt-huang** [comment]
@geyuxiao The delay between tf.estimator.train_and tf.estimator.evalute depends on your model and datasets, among other factors.
The delay is usually very small so the difference in loss should be minimal and is unlikely to have statistical and practical significance.

**Comment by wt-huang** [comment]
Closing as this resolved, feel free to reopen if problem persists

**Comment by geyuxiao** [comment]
There are no errors, but please note the parameters here.

**Comment by hxs91** [comment]
The same problem occurs for me.
For `train_and_evaluate`, I have the following codes:
```
    train_input_fn = input_fn_builder(True, cfg, train_examples, label_to_id, id_to_label, vocab)
    dev_examples, _ = processor.get_dev_examples(FLAGS.data_dir)
    eval_input_fn = input_fn_builder(False, cfg, dev_examples, label_to_id, id_to_label, vocab)
    train_spec = tf.estimator.TrainSpec(input_fn=train_input_fn, max_steps=num_train_steps)
    eval_spec = tf.estimator.EvalSpec(input_fn=eval_input_fn, throttle_secs=60)
    tf.estimator.train_and_evaluate(estimator, train_spec, eval_spec)
```
and I get the log as follows:
`Saving dict for global step 223632: acc = 0.92796874, global_step = 223632, loss = 52.85372`
For `estimator.evaluate`, I use the same model trained during `train_and_evaluate`, the codes are as follows:
```
    dev_examples, _ = processor.get_dev_examples(FLAGS.data_dir)
    eval_input_fn = input_fn_builder(False, cfg, dev_examples, label_to_id, id_to_label, vocab)
    result = estimator.evaluate(input_fn=eval_input_fn)
```
But the log gives a different and poor result:
`Saving dict for global step 223632: acc = 0.90173215, global_step = 223632, loss = 73.50448`
This really makes me confused. @wt-huang

**Comment by hxs91** [comment]
update:
I find the difference is due to the different parameter `steps` in `train_and_evaluate` and `estimator.evaluate`.
`steps` in `train_and_evaluate` has a default value of `100`, but the default value of it in `estimator.evaluate` is `None`.
After I set it to `100`, this problem is resolved. But I still do not know the reason, and I wonder whether whether this result is identical to the `estimator.predict` or not.

**Comment by hxs91** [comment]
It seems that `steps` in `estimator.evaluate` controls the proportion of evaluation data, but what exactly does `steps` mean?

**Comment by hxs91** [comment]
> It seems that `steps` in `estimator.evaluate` controls the proportion of evaluation data, but what exactly does `steps` mean?

Sorry for this stupid question. I know it now.
one `step` in this situation means an evaluation for a batch of data.

**Comment by rocklee2013** [comment]
> There are no errors, but please note the parameters here.
> 
> `eval_spec = tf.estimator.EvalSpec( input_fn=lambda: input_fn.read_fn(validation_data, steps=None )`
> 
> steps = None represent all test sample data.
> If you don't set it, tf defaults to 100.
> I did not set this parameter before, resulting in 100 samples being evaluated each time, so it is not compatible with the overall sample. So not the delay you mentioned @wt-huang .
> Anyway, thank you for your previous reply.

hi , I have a quetion. "resulting in 100 samples being evaluated each time" , here "100 samples " means 100*batchSize samples or just 100  samples ?

**Comment by mhajiaghayi** [comment]
I had faced a similar case where the train and evaluate results were so different. it turned out that it had to do with batch normalization in my code.

**Comment by mohapatras** [comment]
@mhajiaghayi Can you please suggest the solution for fixing the batchnormalization issue?

**Comment by ZionWong** [comment]
> I had faced a similar case where the train and evaluate results were so different. it turned out that it had to do with batch normalization in my code.

I met the same problem, have you fixed it?
